### PR TITLE
Add settings preview to editor dark mode

### LINKED ISSUE #2570: Addon settings preview
Related: #475

### Is your request related to a problem? Please describe.

Editor dark mode allows the user to customize colors, but it isn't always obvious which color does what. Website dark mode will have a similar feature in the future.

### Describe the solution you'd like

An option in `addon.json` to specify a Vue component displayed on the settings page as a preview of the settings. Changing a setting would update the preview, and clicking an area in the preview would highlight the setting that affects it.

### PR DESCRIPTION
**Pull request by mxmou**
Resolves #2570

### Changes

Adds a preview to the editor dark mode addon, displayed above the settings.
* You can switch tabs and toggle full screen.
* Moving the mouse over a setting highlights its affected area in the preview.
* Moving the mouse over a setting's affected area highlights the setting.

image

Each preset has a small "palette" showing the most important colors.

image

### CODE REVIEW

**Comment by LankyBox01** [comment]
oooooooo time to test locally

**Comment by LankyBox01** [comment]
image
(slightly zoomed out)

works perfectly

**Comment by WorldLanguages** [comment]
mxmou king 👑

**Comment by mxmou** [comment]
Is there a way to hide the warnings about style interpolation and Internet Explorer?

**Comment by mxmou** [comment]
This adds two new properties to `addon.json`, `addonPreview` and `presetPreview`, which are both objects with a `type` property that has to be set to the component name without the `preview-` prefix. They can also contain other properties that can be accessed by the component.

File: webpages/settings/index.js [hunk 1]
```diff
@@ -48,6 +48,8 @@ chrome.storage.sync.get(["globalTheme"], function ({ globalTheme = false }) {
     "webpages/settings/components/addon-group-header",
     "webpages/settings/components/addon-body",
     "webpages/settings/components/category-selector",
+    "webpages/settings/components/previews/editor-dark-mode",
+    "webpages/settings/components/previews/palette",
```
Comment (WorldLanguages) [inline webpages/settings/index.js:52]: These need to be added to preload in \<head>.

**Comment by WorldLanguages** [comment]
> You can switch tabs and toggle full screen.

Could we make this more noticeable? I didn't notice we could switch tabs until I saw the cursor change when hovering the full screen button. At a minimum, I think code/costumes/sounds tabs should have `cursor: pointer;`

**Comment by WorldLanguages** [comment]
I don't know how we could solve the issue, but in most screens it's necessary to zoom out or scroll up and down constantly to properly experience the preview with the settings located closest to the bottom.

**Comment by WorldLanguages** [comment]
Other than that, this PR is fantastic lol
@mxmou What's the next step? Exporting single addon settings as JSON? And letting users submit to us? Then we can have a real "themes" tabs with user made presets or superpresets :D

**Comment by mxmou** [comment]
> > You can switch tabs and toggle full screen.
> 
> Could we make this more noticeable? I didn't notice we could switch tabs until I saw the cursor change when hovering the full screen button. At a minimum, I think code/costumes/sounds tabs should have `cursor: pointer;`

You're right, the tabs should have `cursor: pointer`. I'm not sure if there's a better way to indicate that they're clickable: the scale is too small for e.g. replacing the text and icon placeholders on clickable elements with actual text.

**Comment by mxmou** [comment]
> I don't know how we could solve the issue, but in most screens it's necessary to zoom out or scroll up and down constantly to properly experience the preview with the settings located closest to the bottom.

Making the preview sticky might work.

**Comment by mxmou** [comment]
> @mxmou What's the next step? Exporting single addon settings as JSON? And letting users submit to us? Then we can have a real "themes" tabs with user made presets or superpresets :D

I think a more important feature is being able to save your current settings for an addon to an editable preset. Then the ability to export presets could be added.

**Comment by WorldLanguages** [comment]
@mxmou It's up to you to decide when this gets merged, as long as the preload comment I made above is resolved and `cursor: pointer` is added to tabs

**Comment by WorldLanguages** [comment]
mxmou sorry if I messed up with your planning by constantly changing the milestone dates lol. But v1.20.0 will be released in the following days. This can get in if the preview gets a little smaller, preload is fixed and tabs get pointer cursor

**Comment by mxmou** [comment]
> Is there a way to hide the warnings about style interpolation and Internet Explorer?

I meant these:
image

`v-bind:style` doesn't support CSS variables, at least in this version of Vue, so that isn't a solution.

**Comment by mxmou** [comment]
> This can get in if the preview gets a little smaller, preload is fixed and tabs get pointer cursor

Done.

**Comment by WorldLanguages** [comment]
This won't get into v1.20.0

**Comment by WorldLanguages** [comment]
> I meant these:

Ohhh, I see what you mean now.

**Comment by WorldLanguages** [comment]
Hmm I think this PR broke color pickers. Nothing happens when I click on them.

**Comment by WorldLanguages** [comment]
To avoid the Internet Explorer warnings, I thought of two alternatives.

Number 1 is to set the variables from JavaScript:
```js
// editor-dark-mode.js
    methods: {
      selectTab(id) {
        this.selectedTab = id;
        this.$emit("areahover", "activeTab");
      },
      toggleFullScreenView() {
        this.fullScreenView = !this.fullScreenView;
      },
      setCssVariables() {
        const vars = {
          page: this.settings.page,
          "page-text": this.colors.pageText,
          primary: this.settings.primary,
          "primary-transparent": this.colors.primaryTransparent,
          "primary-text": this.colors.primaryText,
          highlightText: this.settings.highlightText,
          menuBar: this.settings.menuBar,
          "menuBar-text": this.colors.menuBarText,
          "menuBar-border": this.colors.menuBarBorder,
          accent: this.settings.accent,
          "accent-text": this.colors.accentText,
          "accent-transparentText": this.colors.accentTransparentText,
          "accent-artboard": this.colors.accentArtboard,
          "accent-checkerboard": this.colors.accentCheckerboard,
          input: this.settings.input,
          "input-transparent": this.colors.inputTransparent,
          "input-text": this.colors.inputText,
          tab: this.settings.tab,
          "tab-text": this.colors.tabText,
          activeTab: this.settings.activeTab,
          workspace: this.settings.workspace,
          categoryMenu: this.settings.categoryMenu,
          "categoryMenu-text": this.colors.categoryMenuText,
          "categoryMenu-selection": this.colors.categoryMenuSelection,
          palette: this.colors.palette,
          selector: this.settings.selector,
          "selector-text": this.colors.selectorText,
          selector2: this.settings.selector2,
          "selector2-text": this.colors.selector2Text,
          selectorSelection: this.settings.selectorSelection,
          fullscreen: this.settings.fullscreen,
          stageHeader: this.settings.stageHeader,
          border: this.hoveredSettingId === "border" ? 'var("brand-orange)' : this.settings.border,
        };

        for (const key of Object.keys(vars)) {
          this.$el.style.setProperty(`--${key}`, vars[key]);
        }
      },
    },
    watch: {
      settings() {
        this.setCssVariables();
      },
      colors() {
        this.setCssVariables();
      },
      hoveredSettingId() {
        this.setCssVariables();
      },
    },
```

Number 2 is to trap console.error:
```js
const originalConsoleError = console.error;
console.error = function () {
  if (
    arguments[0] &&
    typeof arguments[0] === "string" &&
    arguments[0].includes(
      'interpolation in "style" attribute will cause the attribute to be discarded in Internet Explorer'
    )
  )
    return;
  return originalConsoleError.apply(console, arguments);
};
```

**Comment by apple502j** [comment]
Solution 1 seems better.

**Review by WorldLanguages — APPROVED** [review]

**Pull request merged**